### ISSUE TITLE
[FR] Configurable stance for Pydantic's `extra=ignore/allow`

### ISSUE THREAD

**Issue by tattyd** [astral-sh/ty#3966]
### Summary

Hi there!

Overjoyed to see the Pydantic updates earlier today! I've been waiting months to finally integrate ty into our pydantic-heavy codebase :)

`ty` (rightly) respects `extra=ignore/allow`, and when `extra=ignore` it won't flag any type errors for unknown constructor parameters. `mypy`, however, still checks constructor params regardless of the `extra` setting.

Whilst ty's behaviour is correct by the book, in reality we've found what we really want is to flag unknown parameters regardless - the extra=ignore is often for compatibility with legacy data at rest, and there's little valid use case (I think?) to pass in unknown params that get dropped on the floor.

I'm wondering if it's possible to grow a configuration option for `ty`, something like `pydantic_still_validate_init_params_even_if_extra_is_ignored` (bad name but hopefully you get it :) )

I had Claude whip up a repro script you should be able to paste into your shell, below.

Thanks for all your work!
David

## Repro

```sh
mkdir ty-extra-repro && cd ty-extra-repro
uv venv && uv pip install pydantic mypy
cat > repro.py <<'EOF'
from pydantic import BaseModel, ConfigDict


class User(BaseModel):  # pydantic default: extra="ignore"
    name: str


class StrictUser(BaseModel):
    model_config = ConfigDict(extra="forbid")
    name: str


User(name="a", nmae="typo")        # ty: not flagged — but 'nmae' is silently dropped at runtime
StrictUser(name="a", nmae="typo")  # ty: flagged (unknown-argument)
EOF
cat > mypy.ini <<'EOF'
[mypy]
plugins = pydantic.mypy

[pydantic-mypy]
init_forbid_extra = True
EOF
uvx ty check repro.py --python .venv
.venv/bin/mypy repro.py
.venv/bin/python -c "
from pydantic import BaseModel
class User(BaseModel):
    name: str
print(User(name='a', nmae='typo'))  # prints name='a' — typo silently dropped, no error
"
```

## Observed (ty 0.0.58, mypy 2.2.0, pydantic 2.13.4)

ty — flags only the `extra="forbid"` model:

```
error[unknown-argument]: Argument `nmae` does not match any known parameter
  --> repro.py:14:22
Found 1 diagnostic
```

mypy with the pydantic plugin's `init_forbid_extra = True` — flags both:

```
repro.py:13: error: Unexpected keyword argument "nmae" for "User"  [call-arg]
repro.py:14: error: Unexpected keyword argument "nmae" for "StrictUser"  [call-arg]
Found 2 errors in 1 file (checked 1 source file)
```

Runtime — the typo is silently swallowed on the default-config model:

```
name='a'
```

### Version

0.0.58

**Comment by sharkdp** [comment]
> Overjoyed to see the Pydantic updates earlier today! I've been waiting months to finally integrate ty into our pydantic-heavy codebase :)

Thank you for the feedback!



> `ty` (rightly) respects `extra=ignore/allow`, and when `extra=ignore` it won't flag any type errors for unknown constructor parameters. `mypy`, however, still checks constructor params regardless of the `extra` setting.
> 
> Whilst ty's behaviour is correct by the book, in reality we've found what we really want is to flag unknown parameters regardless - the extra=ignore is often for compatibility with legacy data at rest, and there's little valid use case (I think?) to pass in unknown params that get dropped on the floor.
> 
> I'm wondering if it's possible to grow a configuration option for `ty`, something like `pydantic_still_validate_init_params_even_if_extra_is_ignored`

This makes sense to me. Instead of adding a configuration setting, I'm thinking about adding a dedicated lint rule that would be enabled by default (with "warning" severity). Users that do not want that strict behavior could then easily opt out by disabling that specific rule. What do you think?

**Comment by tattyd** [comment]
> Users that do not want that strict behavior could then easily opt out by disabling that specific rule. What do you think?

Sounds great!
